Blog: Native Sidecars (#13632)
istio/istio.io: content/en/blog/2023/native-sidecars/index.md

---
title: Kubernetes Native Sidecars in Istio
description: Demoing the new SidecarContainers feature with Istio.
publishdate: 2023-07-25
attribution: "John Howard (Google)"
keywords: [istio,sidecars,kubernetes]
---

> From the beginning of Istio, applications have been augmented with service mesh functionality by using the "sidecar" pattern:

Reviewer: nit, it is a good practice to tell what this blog post about in a first sentence. So snippets are better extracted.  Something along the lines:

From the beginning of Istio... sidecar pattern. Kubernetes is winning new workloads and Istio hits limitations of sidecar implementation. With 1.28 and native sidecar support, Istio can be used safely and more reliably on these new workloads.

Reviewer 2: I would go with something like

"If you have heard anything about service meshes, it is that they work using the _sidecar pattern_: a proxy server is deployed alongside your application code.  The sidecar pattern is just that: a pattern; up until this point, there has been no formal support for 'sidecar containers' in Kubernetes at all.  This has caused no end of problems: what if you have a job that terminates by design, but a sidecar container that doesn't?  This exact use case is the [most upvoted issue ever on the Kubernetes GitHub](https://github.com/kubernetes/kubernetes/issues/25908).

A formal proposal for adding sidecar support in Kubernetes was raised in 2019, with many stops and starts along the way. After a reboot of the project last year, formal support for sidecars is being released to Alpha in Kubernetes 1.28. Istio has implemented support for this feature, and in this post you can learn how to update your configuration to take advantage of it."

Author: I lifted most of Craig's comment - thanks!

each Pod gets an additional container running alongside it.
In Istio's case, this additional container runs the Istio proxy, giving us the classic Istio architecture:

{{< image width="80%"
    link="/docs/ops/deployment/architecture/arch.svg"
    caption="Istio Architecture"
    >}}

While most common in service meshes, the sidecar pattern is used for other contexts as well, such as monitoring agents.

## Sidecar woes

Sidecar containers give a lot of power, but come with some issues.
While containers within a Pod can share some things, their *lifecycle's* are entirely decoupled.
To Kubernetes, both of these containers are functionally the same.

However, in Istio they are not the same - the Istio container is required for the primary application container to run,

> and has no value without the primary application container.

Reviewer: I'd say the opposite sometimes also true - when sidecar provides secure connection, primary container doesn;t make sense without the sidecar as it cannot communicate. Thus we will restart siecar even when Job is finite.

Author: True - I guess neither makes sense without the other :-) I was thinking the sidecar does nothing on its own though, it only serves to augment the main app. let me reword this

* If an application container intentionally exists (typically from usage in a `Job`), Istio's container will still run and keep the pod running - indefinitely.
  This is also a [top GitHub issue](https://github.com/istio/istio/issues/11659).
* `InitContainers`, which run before Istio's container starts, cannot access the network.

Countless hours have been spent in the Istio community and beyond to work around these - to limited success.

## Fixing the root cause

While increasingly-complex workarounds in Istio can help alleviate the pain for Istio users, ideally all of this would just work - and not just for Istio.
Fortunately, the Kubernetes community has been hard at work to address these directly in Kubernetes.

In Kubernetes 1.28, a new feature to add native support for sidecars was merged, closing out over 5 years of ongoing work.
With this merged, all of our issues can be addressed without workarounds!

While we are on the "GitHub issue hall of fame", [these](https://github.com/kubernetes/kubernetes/issues/25908) [issues](https://github.com/kubernetes/kubernetes/issues/65502) account for #1 and #6 all time issues in Kubernetes - and have finally been closed!

> A special thanks goes to the huge group of individuals involved in getting this past the finish line.

Author: @SergeyKanzhelev  @matthyx - is there a bounded list of people we can give a shout-out too? I wanted to include individuals but didn't want to leave folks off so left it generic. LMK if you have thoughts/preferences/etc

Reviewer: I can provide a list of ~10-20 people who was most involved recently. Unfortunately I don't remember everybody involved in early incarnations of this proposal. Maybe it will be better to make this posted separately as a blog post on kubernetes.dev. Let me think about it.

Reviewer 3: Thanks for doing this blog post, I agree with Sergey since the final result is the continuation of so many people work. Let's not upset anyone :-)

## Trying it out

While Kubernetes 1.28 is not out yet, and the feature is Alpha (and therefor, off by default), and the support for the feature in Istio is not yet shipped, we can still try it out today - just don't try this in prod!

First, we need to spin up a pre-release Kubernetes cluster, with the `SidecarContainers` feature enabled:

{{< text shell >}}
$ cat <<EOF | kind create cluster --name sidecars --image gcr.io/istio-testing/kind-node:v1.28.0-beta.0 --config=-
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
featureGates:
  SidecarContainers: true
EOF
{{< /text >}}

Then we can download the latest Istio pre-release. I used Linux here; see [Dev Builds](https://github.com/istio/istio/wiki/dev-builds) for more steps.
These images are built from the latest Istio commit, so again - do not try in prod!

> When we install we will enable the feature flag for native sidecar support, and turn on access logs to help demo things later.

Reviewer 4: ```suggestion
When we install, we will enable the feature flag for native sidecar support and turn on access logs to help demo things later.
```

> Then we can download the latest Istio pre-release. I used Linux here; see [Dev Builds](https://github.com/istio/istio/wiki/dev-builds) for more steps.

Reviewer 5: Is it possible we link the related istio PR in this blog post? Then user will know for sure which build they should use for this feature.

Author: Its like 8 PRs. Users should use either HEAD or 1.19+

{{< /text >}}

Let's look at the pod:

{{< text shell >}}
$ kubectl get pods
NAME                     READY   STATUS    RESTARTS   AGE
sleep-7656cf8794-8fhdk   2/2     Running   0          51s
{{< /text >}}

Everything looks normal at first glance...
If we look under the hood, we can see the magic, though.

{{< text shell >}}
$ kubectl get pod -o "custom-columns="\
"NAME:.metadata.name,"\
"INIT:.spec.initContainers[*].name,"\
"CONTAINERS:.spec.containers[*].name"

NAME                     INIT                     CONTAINERS
sleep-7656cf8794-8fhdk   istio-init,istio-proxy   sleep
{{< /text >}}

Here we can see all the `containers` and `initContainers` in the pod.

Surprise! `istio-proxy` is now an `initContainer`.

More specifically, it is an `initContainer` with `restartPolicy: Always` set (a new field, enabled by the `SidecarContainers` feature).
This tells Kubernetes to treat it as a sidecar.

This means that later containers in the list of `initContainers`, and all normal `containers` will not start until proxy container is first ready.
Additionally, the pod will terminate even if the proxy container is still running.

### Init container traffic

> To put this to the test, lets make our pod actually do something.

Reviewer 4: ```suggestion
To put this to the test, let's make our pod actually do something.
```

Here we deploy a simple pod that sends a request in an `initContainer`.
Normally, this would fail.

{{< text yaml >}}
apiVersion: v1
kind: Pod
metadata:
  name: sleep
spec:
  initContainers:
  - name: check-traffic
    image: istio/base
    command:
    - curl
    - httpbin.org/get
  containers:
  - name: sleep
    image: istio/base
    command: ["/bin/sleep", "infinity"]
{{< /text >}}

Checking the proxy container, we can see the request both succeeded and went through the Istio sidecar:

{{< text shell >}}
$ kubectl logs sleep -c istio-proxy | tail -n1
[2023-07-25T22:00:45.703Z] "GET /get HTTP/1.1" 200 - via_upstream - "-" 0 1193 334 334 "-" "curl/7.81.0" "1854226d-41ec-445c-b542-9e43861b5331" "httpbin.org" ...
{{< /text >}}

### Exiting pods

> Earlier we mentioned that when applications exit (common in `Jobs`), the Pod would live forever.

Reviewer 4: ```suggestion
Earlier, we mentioned that when applications exit (common in `Jobs`), the Pod would live forever.
```

Fortunately, this is addressed as well!

First we deploy a pod that will exit after one second and doesn't restart:

{{< text yaml >}}
apiVersion: v1
kind: Pod
metadata:
  name: sleep
spec:
  restartPolicy: Never
  containers:
- name: sleep
  image: istio/base
  command: ["/bin/sleep", "1"]
{{< /text >}}

And we can watch its progress:

{{< text shell >}}
$ kubectl get pods -w
NAME    READY   STATUS     RESTARTS   AGE
sleep   0/2     Init:1/2   0          2s
sleep   0/2     PodInitializing   0          2s
sleep   1/2     PodInitializing   0          3s
sleep   2/2     Running           0          4s
sleep   1/2     Completed         0          5s
sleep   0/2     Completed         0          12s
{{< /text >}}

Here we can see the application container exited, and shortly after Istio's sidecar container exits as well.

> No more zombie pods!

Reviewer: if you can add a scenario when you kill the istio and shows it being restarted, it will be a good demo too

Reviewer: or set the very low limits so it will be OOM killed and restarted.

## What about ambient mode?

Last year, Istio announced [ambient mode](/blog/2022/introducing-ambient-mesh/) - a new data plane mode for Istio that doesn't rely on sidecar containers.
So with that coming, does any of this even matter?

I would say a resounding "Yes"!

While the impacts of sidecar are lessened when ambient mode is used for a workload, I expect that almost all large scale Kubernetes users have some sort of sidecar in their deployments.

> This could be Istio workloads they don't want to migrate to ambient, that they haven't *yet* migrated, or things unrelated to Istio.

Reviewer: can it be Istio with the very "special" configuration of an envoy that ambient cannot provide? I don't know enough to suggest any specifics, but from monitoring background - whenever we say ambient or central, all customizations are being lost immediately.

Author: Yeah that is one of the primary cases we expect them to be kept (all theoretical, since we haven't properly launched it yet to get real feedback!)

> So while there may be a less scenarios that this matters, it still is a huge improvement for the cases where sidecars are used.

Reviewer 4: ```suggestion
So while there may be less scenarios where this matters, it still is a huge improvement for the cases where sidecars are used.
```

You may wonder the opposite - if all our sidecar woes are addressed, why do we need ambient mode at all?
There are still a variety of benefits ambient brings with these sidecar limitations addressed.
For example, [this blog post](/blog/2023/waypoint-proxy-made-simple/) goes into details about why decoupling proxies from workloads is advantageous, and [this post](/blog/2023/ambient-performance-savings/) covers some of the cost saving opportunities.

## Try it out yourself

We encourage the adventurous readers to try this out themselves in testing environments!

> Feedback for these experimental and alpha features is critical to ensure they are stable and meeting expectations before shipping broader.

Reviewer: we are especially looking for the feedback on 
- the shutdown sequence and multiple sidecars shutdown sequence.
- backoff timeout when envoy is crashing. Anything we can do wrt readiness of a pod or liveness of a Pod when sidecars present
- any other scenarios we forgot

Reviewer: would be nice it you can list the feedback items explicitly

---
title: Kubernetes Native Sidecars in Istio
description: Demoing the new SidecarContainers feature with Istio.
publishdate: 2023-07-25
attribution: "John Howard (Google)"
keywords: [istio,sidecars,kubernetes]
---

If you have heard anything about service meshes, it is that they work using the sidecar pattern: a proxy server is deployed alongside your application code.
The sidecar pattern is just that: a pattern.

> Up until this point, there has been no formal support for 'sidecar containers' in Kubernetes at all.

Reviewer 6: I'm not sure on the single quotes and how they look. It doesn't make sense to backtick them. Maybe just remove them.

> This has caused no end of problems: what if you have a job that terminates by design, but a sidecar container that doesn't?

Reviewer 6: This can be rewriten differently leaving `no end` but maybe just remove it
```suggestion
This has caused a number of problems: what if you have a job that terminates by design, but a sidecar container that doesn't?
```

This exact use case is the [most popular ever on the Kubernetes issue tracker](https://github.com/kubernetes/kubernetes/issues/25908).

> A formal proposal for adding sidecar support in Kubernetes was raised in 2019, with many stops and starts along the way.
> After a reboot of the project last year, formal support for sidecars is being released to Alpha in Kubernetes 1.28.

Reviewer 6: ```suggestion
A formal proposal for adding sidecar support in Kubernetes was raised in 2019. With many stops and starts along the way,
and after a reboot of the project last year, formal support for sidecars is being released to Alpha in Kubernetes 1.28.
```

Istio has implemented support for this feature, and in this post you can learn how to take advantage of it.

## Sidecar woes

Sidecar containers give a lot of power, but come with some issues.

> While containers within a Pod can share some things, their *lifecycle's* are entirely decoupled.

Reviewer 6: We don't really capitalize Pod anywhere else:
```suggestion
While containers within a pod can share some things, their *lifecycle's* are entirely decoupled.
```

To Kubernetes, both of these containers are functionally the same.

However, in Istio they are not the same - the Istio container is required for the primary application container to run,
and has no value without the primary application container.

This mismatch in expectation leads to a variety of issues:
* If the application container starts faster than Istio's container, it cannot access the network.

>   This wins the [most +1's](https://github.com/istio/istio/issues/11130) on GitHub by a landslide.

Reviewer 6: ```suggestion
  This wins the [most +1's](https://github.com/istio/istio/issues/11130) on Istio's GitHub by a landslide.
```

> * If Istio's container shuts down before the application container, it cannot access the network.

Reviewer 6: ```suggestion
* If Istio's container shuts down before the application container, the application container cannot access the network.
```

>   This often means we shutdown too fast (breaking the network) or too slow (leaving pods running with no purpose).

Reviewer 6: I might think about removing this sentence. It's just a recap of the first two items and sort of separates the rest from this list.

* If an application container intentionally exits (typically from usage in a `Job`), Istio's container will still run and keep the pod running indefinitely.
  This is also a [top GitHub issue](https://github.com/istio/istio/issues/11659).
* `InitContainers`, which run before Istio's container starts, cannot access the network.

> Countless hours have been spent in the Istio community and beyond to work around these - to limited success.

Reviewer 6: ```suggestion
Countless hours have been spent in the Istio community and beyond to work around these issues - to limited success.
```

## Fixing the root cause

While increasingly-complex workarounds in Istio can help alleviate the pain for Istio users, ideally all of this would just work - and not just for Istio.
Fortunately, the Kubernetes community has been hard at work to address these directly in Kubernetes.

In Kubernetes 1.28, a new feature to add native support for sidecars was merged, closing out over 5 years of ongoing work.
With this merged, all of our issues can be addressed without workarounds!

> While we are on the "GitHub issue hall of fame", [these](https://github.com/kubernetes/kubernetes/issues/25908) [issues](https://github.com/kubernetes/kubernetes/issues/65502) account for #1 and #6 all time issues in Kubernetes - and have finally been closed!

Reviewer 6: I'm not sure of a better way to write this, but looking at in initially I was wondering where the second issue was (it looks like a single link). Maybe add the word two in between:
```suggestion
While we are on the "GitHub issue hall of fame", [these](https://github.com/kubernetes/kubernetes/issues/25908) two [issues](https://github.com/kubernetes/kubernetes/issues/65502) account for #1 and #6 all time issues in Kubernetes - and have finally been closed!
```

A special thanks goes to the huge group of individuals involved in getting this past the finish line.

## Trying it out

> While Kubernetes 1.28 is not out yet, and the feature is Alpha (and therefore, off by default), and the support for the feature in Istio is not yet shipped, we can still try it out today - just don't try this in prod!

Reviewer 6: It is today.

> First, we need to spin up a pre-release Kubernetes cluster, with the `SidecarContainers` feature enabled:

Reviewer 6: Do we still want to use pre-release? Need to also tweak the command line.
```suggestion
First, we need to spin up a Kubernetes 1.28 cluster, with the `SidecarContainers` feature enabled:
```

{{< text shell >}}
$ cat <<EOF | kind create cluster --name sidecars --image gcr.io/istio-testing/kind-node:v1.28.0-beta.0 --config=-
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
featureGates:
  SidecarContainers: true
EOF
{{< /text >}}

Then we can download the latest Istio pre-release. I used Linux here; see [Dev Builds](https://github.com/istio/istio/wiki/dev-builds) for more steps.

> These images are built from the latest Istio commit, so again - do not try in prod!

Reviewer 6: ```suggestion
These images are built from the latest Istio commit, so again - do not try this in prod!
```

Reviewer 6: Maybe we want to expand prod -> production here and in the comment above. (and elsewhere).

> When we install, we will enable the feature flag for native sidecar support and turn on access logs to help demo things later.

Reviewer 6: ```suggestion
When we install Istio, we will enable the feature flag for native sidecar support and turn on access logs to help demo things later.
```

{{< /text >}}

And finally we can deploy a workload:

{{< text shell >}}
$ kubectl label namespace default istio-injection=enabled
$ kubectl apply -f samples/sleep/sleep.yaml
{{< /text >}}

Let's look at the pod:

{{< text shell >}}
$ kubectl get pods
NAME                     READY   STATUS    RESTARTS   AGE
sleep-7656cf8794-8fhdk   2/2     Running   0          51s
{{< /text >}}

Everything looks normal at first glance...
If we look under the hood, we can see the magic, though.

{{< text shell >}}
$ kubectl get pod -o "custom-columns="\
"NAME:.metadata.name,"\
"INIT:.spec.initContainers[*].name,"\
"CONTAINERS:.spec.containers[*].name"

NAME                     INIT                     CONTAINERS
sleep-7656cf8794-8fhdk   istio-init,istio-proxy   sleep
{{< /text >}}

Here we can see all the `containers` and `initContainers` in the pod.

Surprise! `istio-proxy` is now an `initContainer`.

More specifically, it is an `initContainer` with `restartPolicy: Always` set (a new field, enabled by the `SidecarContainers` feature).
This tells Kubernetes to treat it as a sidecar.

> This means that later containers in the list of `initContainers`, and all normal `containers` will not start until proxy container is first ready.

Reviewer 6: I'm not sure if we want to say proxy or istio-proxy.
```suggestion
This means that later containers in the list of `initContainers`, and all normal `containers` will not start until the proxy container is ready.
```

Additionally, the pod will terminate even if the proxy container is still running.

### Init container traffic

To put this to the test, let's make our pod actually do something.
Here we deploy a simple pod that sends a request in an `initContainer`.
Normally, this would fail.

> {{< text yaml >}}

Reviewer 7: Can you use the deploy command and mention this will have istio-proxy sidecar injected?  I assume this is to the default ns that has injection label.

apiVersion: v1
kind: Pod
metadata:
  name: sleep
spec:
  initContainers:
  - name: check-traffic
    image: istio/base
    command:
    - curl
    - httpbin.org/get
  containers:
  - name: sleep
    image: istio/base
    command: ["/bin/sleep", "infinity"]
{{< /text >}}

> Checking the proxy container, we can see the request both succeeded and went through the Istio sidecar:

Reviewer 6: Do we want to say something about how istio-proxy ends up injected before check-traffic? I assume if it doesn't, it would fail like before.

Reviewer 7: Yes worth clarifying. I had the same q when i read it.  The sidecar should be injected & running (to my understanding) before checking the traffic, while previously without the new sidecar feature from k8s, the culr httpbin.org would likely run before istio-proxy reaches running thus failed.

{{< text shell >}}
$ kubectl logs sleep -c istio-proxy | tail -n1
[2023-07-25T22:00:45.703Z] "GET /get HTTP/1.1" 200 - via_upstream - "-" 0 1193 334 334 "-" "curl/7.81.0" "1854226d-41ec-445c-b542-9e43861b5331" "httpbin.org" ...
{{< /text >}}

### Exiting pods

Earlier, we mentioned that when applications exit (common in `Jobs`), the Pod would live forever.
Fortunately, this is addressed as well!

First we deploy a pod that will exit after one second and doesn't restart:

> 

Reviewer 7: Suggest using the deploy command and highlight this is deployed to the default ns that has injection enabled.

{{< text yaml >}}
apiVersion: v1
kind: Pod
metadata:
  name: sleep
spec:
  restartPolicy: Never
  containers:
- name: sleep
  image: istio/base
  command: ["/bin/sleep", "1"]
{{< /text >}}

And we can watch its progress:

{{< text shell >}}
$ kubectl get pods -w
NAME    READY   STATUS     RESTARTS   AGE
sleep   0/2     Init:1/2   0          2s
sleep   0/2     PodInitializing   0          2s
sleep   1/2     PodInitializing   0          3s
sleep   2/2     Running           0          4s
sleep   1/2     Completed         0          5s
sleep   0/2     Completed         0          12s
{{< /text >}}

> Here we can see the application container exited, and shortly after Istio's sidecar container exits as well.

Reviewer 7: Unclear how to tell this from the output, could use a little more explanation.

No more zombie pods!

## What about ambient mode?

Last year, Istio announced [ambient mode](/blog/2022/introducing-ambient-mesh/) - a new data plane mode for Istio that doesn't rely on sidecar containers.

> So with that coming, does any of this even matter?

Reviewer 6: ```suggestion
So with ambient mode coming, does any of this even matter?
```

A special thanks goes to the huge group of individuals involved in getting this past the finish line.

## Trying it out

While Kubernetes 1.28 was just released, the new `SidecarContainers` feature is Alpha (and therefore, off by default), and the support for the feature in Istio is not yet shipped, we can still try it out today - just don't try this in production!

First, we need to spin up a Kubernetes 1.28 cluster, with the `SidecarContainers` feature enabled:

{{< text shell >}}
$ cat <<EOF | kind create cluster --name sidecars --image gcr.io/istio-testing/kind-node:v1.28.0 --config=-
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
featureGates:
  SidecarContainers: true
EOF
{{< /text >}}

Then we can download the latest Istio 1.19 pre-release (as 1.19 is not yet out). I used Linux here.
This is a pre-release of Istio, so again - do not try this in production!
When we install Istio, we will enable the feature flag for native sidecar support and turn on access logs to help demo things later.

{{< text shell >}}
$ TAG=1.19.0-beta.0

> $ curl -L https://github.com/istio/istio/releases/download/$TAG/istio-$TAG-linux-amd64.tar.gz | tar xz

Reviewer 6: nit: platform is hardcoded. Could adda tip on architecture, but fine for now.
